activerecord/lib/active_record/associations/preloader/through_association.rb

```diff
@@ -74,6 +74,18 @@ def source_preloaders
           end
 
           def middle_records
+            is_middle_records_loaded = true
+            middle_records = owners.each_with_object([]) do |owner, result|
+              association = owner.association(through_reflection.name)
+              if association.loaded?
+                result.concat Array.wrap(association.target) if association.target
+              else
+                is_middle_records_loaded = false
+                break result
+              end
+            end
+            return middle_records if is_middle_records_loaded
+
             through_preloaders.flat_map(&:preloaded_records)
```

Comment:
I think this does the same thing, but is a little bit more readable:

```suggestion 
            middle_records = owners.flat_map do |owner|
              association = owner.association(through_reflection.name)
              break unless association.loaded?

              Array.wrap(association.target)
            end 

            middle_records || through_preloaders.flat_map(&:preloaded_records)
```

Reply:
💯 
Thank you for the suggestion!